crates/sui-core/src/execution_cache/writeback_cache.rs

```diff
@@ -0,0 +1,1158 @@
+struct UncommittedData {
+    /// The object dirty set. All writes go into this table first. After we flush the data to the
+    /// db, the data is removed from this table and inserted into the object_cache.
+    ///
+    /// This table may contain both live and dead objects, since we flush both live and dead
+    /// objects to the db in order to support past object queries on fullnodes.
+    /// When we move data into the object_cache we only retain the live objects.
+    ///
+    /// Further, we only remove objects in FIFO order, which ensures that the the cached
+    /// sequence of objects has no gaps. In other words, if we have versions 4, 8, 13 of
+    /// an object, we can deduce that version 9 does not exist. This also makes child object
+    /// reads efficient. `object_cache` cannot contain a more recent version of an object than
+    /// `objects`, and neither can have any gaps. Therefore if there is any object <= the version
+    /// bound for a child read in objects, it is the correct object to return.
+    objects: DashMap<ObjectID, CachedVersionMap<ObjectEntry>>,
+
+    // Markers for received objects and deleted shared objects. This contains all of the dirty
+    // marker state, which is committed to the db at the same time as other transaction data.
+    // After markers are committed to the db we remove them from this table and insert them into
+    // marker_cache.
+    markers: DashMap<MarkerKey, CachedVersionMap<MarkerValue>>,
+
+    transaction_effects: DashMap<TransactionEffectsDigest, TransactionEffects>,
+
+    // Because TransactionEvents are not unique to the transaction that created them, we must
+    // reference count them in order to know when we can remove them from the cache. For now
+    // we track all referers explicitlyl, but we can use a ref count when we are confident in
+    // the correctness of the code.
+    transaction_events:
+        DashMap<TransactionEventsDigest, (BTreeSet<TransactionDigest>, TransactionEvents)>,
+
+    executed_effects_digests: DashMap<TransactionDigest, TransactionEffectsDigest>,
+
+    // Transaction outputs that have not yet been written to the DB. Items are removed from this
+    // table as they are flushed to the db.
+    pending_transaction_writes: DashMap<TransactionDigest, Arc<TransactionOutputs>>,
+}
+
+impl UncommittedData {
+    fn new() -> Self {
+        Self {
+            objects: DashMap::new(),
+            markers: DashMap::new(),
+            transaction_effects: DashMap::new(),
+            executed_effects_digests: DashMap::new(),
+            pending_transaction_writes: DashMap::new(),
+            transaction_events: DashMap::new(),
+        }
+    }
+}
+
+/// CachedData stores data that has been committed to the db, but is likely to be read soon.
+struct CachedData {
+    // See module level comment for an explanation of caching strategy.
+    object_cache: MokaCache<ObjectID, Arc<Mutex<CachedVersionMap<ObjectEntry>>>>,
```

Comment:
I'm not sure. For mutable objects, contention is impossible (sui already guarantees exclusive access), so Mutex may be fastest.

For read only shared objects and immutable objects, RwLock may be beneficial. But I think when it comes to that we may want an entirely different caching strategy. For instance, it would be good to be able to serve a RO shared obj read without blocking writes to that shared object.

But i'm not even sure any of this matters. The critical sections should be extremely small relative to the duration of transaction execution. All we are doing is a linear search through a small vector followed by an Arc clone to return the Object (or the even cheaper copy needed to return a MarkerValue)

Reply:
let me amend that actually - there can be contention between the write_transaction_outputs and commit_transaction_outputs. but both of these require write access so RwLock won't help there